bittensor/_neuron/text/core_validator/__init__.py

```diff
@@ -606,24 +610,31 @@ def run_epoch( self ):
         self.prometheus_gauges.labels("dividends").set( self.metagraph.dividends[self.uid] )
         self.prometheus_gauges.labels("emission").set( self.metagraph.emission[self.uid] )
 
+        # Iterate epochs.
+        self.epoch += 1
+
     def metagraph_sync(self):
         r""" Syncing metagraph together with other metagraph-size related objects
         """
-        old_hotkeys = self.neuron_hotkeys if self.neuron_hotkeys else self.metagraph.hotkeys
-        self.metagraph.sync()
+        old_hotkeys = self.neuron_hotkeys + [] if self.neuron_hotkeys else self.metagraph.hotkeys
+        self.metagraph.sync(cached=self.config.neuron.metagraph_cached or self.subtensor.network != 'local')
         self.neuron_hotkeys = self.metagraph.hotkeys
 
         changed_hotkeys = []
         # === Reset neuron stats if uid got replaced
         for uid, old_hotkey in enumerate(old_hotkeys):
             if old_hotkey != self.neuron_hotkeys[uid]:
+                block = self.subtensor.block
+                self.neuron_register.setdefault(uid, {})  # [uid] -> dict() of blocks
+                self.neuron_register[uid][block] = {'new_hotkey': self.neuron_hotkeys[uid], 'old_hotkey': old_hotkey}
```

Comment:
This will continuously add more and more memory as the validator runs. I think it's beneficial, primarily to diagnose the network. Should we add a flag, so only validators that we run will store all of this information?

Reply:
Roughly seeing 100KB/day disk space added at current higher rate of deregistrations, so it is not that excessive, but refactored and added a flag `--neuron.track_hotkey_changes` default=False that makes this opt-in.

Reply:
But do we have to store the hotkeys for each of the blocks? I think we can regularly prune the dictionary for hotkeys that is like 2 days ago